diff --git a/Source/JavaScriptCore/dfg/DFGObjectAllocationSinkingPhase.cpp b/Source/JavaScriptCore/dfg/DFGObjectAllocationSinkingPhase.cpp
index 2a1230989344..714f75ea8779 100644
--- a/Source/JavaScriptCore/dfg/DFGObjectAllocationSinkingPhase.cpp
+++ b/Source/JavaScriptCore/dfg/DFGObjectAllocationSinkingPhase.cpp
@@ -846,6 +846,7 @@ private:
                 m_heapAtHead[block].setReached();
                 m_heap = m_heapAtHead[block];
 
+                dataLogLnIf(Options::verboseObjectAllocationSinking(), "LocalHeap when starting ", block, ": ", m_heap);
                 for (Node* node : *block) {
                     handleNode(
                         node,
@@ -1027,7 +1028,8 @@ private:
             // [4] array-allocation-sink-upsilon-with-double-value.js
             Node* base = node->child1().node();
             Allocation* array = m_heap.onlyLocalAllocation(base);
-            if (array && array->isArrayAllocation()) {
+            // We need to check m_sinkCandidates because removing the GetButterfly "allocation" is only valid if the array is sunk too.
+            if (array && array->isArrayAllocation() && m_sinkCandidates.contains(base)) {
                 // Treat GetButterfly as a separate allocation to track its dependency on the Array.
                 // 1. If PutByVal/GetByVal cause escape, materialize both at the same site.
                 // 2. If the Array escapes, also escape the ArrayButterfly since sinking it isn't beneficial.
@@ -1530,6 +1532,8 @@ escapeChildren:
                 if (!m_heap.isAllocation(identifier))
                     m_sinkCandidates.addVoid(identifier);
             }
+
+            dataLogLnIf(Options::verboseObjectAllocationSinking(), "Initial candidates after block ", *block, ": ", listDump(m_sinkCandidates));
         }
 
         auto fixGetButterflyEscapees = [&](auto& escapees) {
@@ -1615,8 +1619,10 @@ escapeChildren:
                     InlineCallFrame* inlineCallFrame = allocation->origin.semantic.inlineCallFrame();
                     if (!inlineCallFrame)
                         continue;
-                    if ((inlineCallFrame->isClosureCall || inlineCallFrame->isVarargs()) && inlineCallFrame != where->origin.semantic.inlineCallFrame())
+                    if ((inlineCallFrame->isClosureCall || inlineCallFrame->isVarargs()) && inlineCallFrame != where->origin.semantic.inlineCallFrame()) {
+                        dataLogLnIf(Options::verboseObjectAllocationSinking(), "Removing candidate because it escapes to a new inline frame: ", allocation);
                         m_sinkCandidates.remove(allocation);
+                    }
                 }
             });
         }
@@ -1626,6 +1632,8 @@ escapeChildren:
         Vector<Node*> worklist;
         worklist.appendRange(m_sinkCandidates.begin(), m_sinkCandidates.end());
 
+        dataLogLn("Candidates after removal: ", listDump(m_sinkCandidates));
+
         while (!worklist.isEmpty()) {
             for (Node* identifier : dependencies.get(worklist.takeLast())) {
                 if (m_sinkCandidates.add(identifier).isNewEntry)
